crates/ruff_linter/src/rules/ruff/rules/sequence_sorting.rs

```diff
@@ -0,0 +1,878 @@
+    // rather than using the expression generator, as this approach allows
+    // us to easily preserve stylistic choices in the original source code
+    // such as whether double or single quotes were used.
+    for (i, (_, elt)) in element_pairs.iter().enumerate() {
+        result.push_str(locator.slice(elt));
+        if i < last_item_index {
+            result.push_str(", ");
+        }
+    }
+
+    result.push_str(closing_paren);
+    result
+}
+
+/// An enumeration of the possible conclusions we could come to
+/// regarding the ordering of the elements in a display of string literals
+#[derive(Debug, is_macro::Is)]
+pub(super) enum SortClassification<'a> {
+    /// It's a display of string literals that is already sorted
+    Sorted,
+    /// It's an unsorted display of string literals,
+    /// but we wouldn't be able to autofix it
+    UnsortedButUnfixable,
+    /// It's an unsorted display of string literals,
+    /// and it's possible we could generate a fix for it
+    UnsortedAndMaybeFixable { items: Vec<&'a str> },
+    /// The display contains one or more items that are not string
+    /// literals.
+    NotAListOfStringLiterals,
+}
+
+impl<'a> SortClassification<'a> {
+    pub(super) fn of_elements(elements: &'a [ast::Expr], sorting_style: &SortingStyle) -> Self {
+        let Some((first, rest @ [_, ..])) = elements.split_first() else {
+            return Self::Sorted;
+        };
+        let Some(string_node) = first.as_string_literal_expr() else {
+            return Self::NotAListOfStringLiterals;
+        };
+        let mut this = string_node.value.to_str();
+
+        for expr in rest {
+            let Some(string_node) = expr.as_string_literal_expr() else {
+                return Self::NotAListOfStringLiterals;
+            };
+            let next = string_node.value.to_str();
+            if sorting_style.compare(next, this).is_lt() {
+                let mut items = Vec::with_capacity(elements.len());
+                for expr in elements {
+                    let Some(string_node) = expr.as_string_literal_expr() else {
+                        return Self::NotAListOfStringLiterals;
+                    };
+                    if string_node.value.is_implicit_concatenated() {
+                        return Self::UnsortedButUnfixable;
+                    }
+                    items.push(string_node.value.to_str());
+                }
+                return Self::UnsortedAndMaybeFixable { items };
+            }
+            this = next;
+        }
+        Self::Sorted
+    }
+}
+
+// An instance of this struct encapsulates an analysis
+/// of a multiline Python tuple/list that represents an
+/// `__all__`/`__slots__`/etc. definition or augmentation.
+pub(super) struct MultilineStringSequenceValue {
+    items: Vec<StringSequenceItem>,
+    range: TextRange,
+    ends_with_trailing_comma: bool,
+}
+
+impl MultilineStringSequenceValue {
+    pub(super) fn num_items(&self) -> usize {
```

Comment:
Nit: I'd expect this to be called `len` (and for this to also implement `is_empty`).

Reply:
Hmm, I tried implementing `is_empty()`, but I get a clippy warning about unused code, because it's never used currently

Reply:
I thought Clippy warned when you had `len` but no `is_empty`, so I must be misremembering!